Manual: new purely syntactic forms

Octachron opened the pull request:
This pull request proposes some documentation for the purely syntactic new features
integrated in 4.03.

Due to the incremental nature of these changes, most of the new documentation
has been added to chapter 6, core language,:
- `_ t` as a short-hand for `(_, ... _) t`. 
- Short functor syntax `S -> T`.
- Type annotations on record fields.
- Type annotations before the `->` in `fun <args> -> <expr>`.

The other changes corresponded to extension of features described in chapter 7, language extension, and thus have been added to this chapter 
- Field punning in object copying expressions `{< x = x; y = y >}`.
- `(type a b)` as syntactic sugar for `(type a) (type b)`.

These changes are still fresh, and I will probably revise them in the upcoming week.

Octachron committed:
Manual: multiple locally abstract types sugar

Octachron committed:
Manual: object copy notation

Octachron committed:
Manual: "_" as anonymous type parameters

Octachron committed:
Manual: functor short syntax

Octachron committed:
Manual: optional type constraint for record label

Octachron committed:
Manual: result type constraint for fun

gasche commented:
This is very good. Could you confirm that merging the PR now is fine? I welcome future changes, but I'm not sure there is value in waiting for them before merging.

Could you maybe add a warning that `fun x y : t -> ...` should not be confused with the very different `fun x (y : t) -> ...`?

Octachron committed:
Manual: result type constraint warning

Octachron commented:
I have added a warning on the syntactic similarity between `fun x y: t ->` and `fun x (y:t) ->` and 
corrected some typographic errors. I am ok with merging this version.
